0.40.0: panic: duplicate metrics collector registration attempted

Coming from https://github.com/NixOS/nixpkgs/pull/292122, where our VM test fails with a similar looking panic from #6510 and #6511.

We start `grafana-agent` via `grafana-agent -config.expand-env -config.file <path to config file>` with the following config:

```yaml
integrations:
  agent:
    enabled: true
    scrape_integration: true
  node_exporter:
    enabled: true
metrics:
  global:
    scrape_interval: 5s
  wal_directory: ${STATE_DIRECTORY}
```

Resulting in the following panic.
Can provide full logs if needed.

```
panic: duplicate metrics collector registration attempted
goroutine 301 [running]:
github.com/prometheus/client_golang/prometheus.(*wrappingRegisterer).MustRegister(0xc002f581b0, {0xc001ec8eb0?, 0x1, 0x3?})
        github.com/prometheus/client_golang@v1.18.0/prometheus/wrap.go:104 +0x145
github.com/prometheus/prometheus/storage/remote.NewWriteHandler({0x89c5200?, 0xc00333bd60}, {0x8a15f10, 0xc002f581b0}, {0x7f8335158a78?, 0xc00333d0a0})
        github.com/prometheus/prometheus@v1.99.0/storage/remote/write_handler.go:55 +0x1ac
github.com/grafana/agent/pkg/metrics/instance.(*Instance).initialize(0xc0033bc1c0, {0x8a3a170, 0xc00333be00}, {0x8a15ee0, 0xc001c59f38}, 0xc0002e2500)
        github.com/grafana/agent/pkg/metrics/instance/instance.go:412 +0x1fa
github.com/grafana/agent/pkg/metrics/instance.(*Instance).Run(0xc0033bc1c0, {0x8a3a170, 0xc00333be00})
        github.com/grafana/agent/pkg/metrics/instance/instance.go:309 +0x328
github.com/grafana/agent/pkg/metrics/instance.(*BasicManager).runProcess(0xc0031d4ac0, {0x8a3a170, 0xc00333be00}, {0xc003335a40, 0x20}, {0x8a58d10, 0xc0033bc1c0})
        github.com/grafana/agent/pkg/metrics/instance/manager.go:265 +0xa3
github.com/grafana/agent/pkg/metrics/instance.(*BasicManager).spawnProcess.func1()
        github.com/grafana/agent/pkg/metrics/instance/manager.go:235 +0x53
created by github.com/grafana/agent/pkg/metrics/instance.(*BasicManager).spawnProcess in goroutine 1
        github.com/grafana/agent/pkg/metrics/instance/manager.go:234 +0x345
```

Version info is different from the official release binaries, because we compile from source in nixpkgs.

```bash
$ grafana-agent --version
agent, version 0.40.0 (branch: v0.40.0, revision: v0.40.0)
  build user:       nix
  build date:       1980-01-01T00:00:00Z
  go version:       go1.21.7
  platform:         linux/amd64
  tags:             builtinassets,nonetwork,nodocker,promtail_journal_enabled
```

Just to avoid confusion:
`ldflags` are set here: https://github.com/NixOS/nixpkgs/blob/bb6e397a6a0318140fb0ba36d7612f1c6704f879/pkgs/servers/monitoring/grafana-agent/default.nix#L34-L44
and build `tags` here: https://github.com/NixOS/nixpkgs/blob/bb6e397a6a0318140fb0ba36d7612f1c6704f879/pkgs/servers/monitoring/grafana-agent/default.nix#L48-L53

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,9 @@ Main (unreleased)
 
 - Fix an issue where Loki could reject a batch of logs when structured metadata feature is used. (@thampiotr)
 
+- Fix a duplicate metrics registration panic when recreating static
+  mode metric instance's write handler. (@rfratto, @hainenber)
+
 v0.40.1 (2024-02-27)
 --------------------
 

--- a/internal/static/metrics/instance/instance.go
+++ b/internal/static/metrics/instance/instance.go
@@ -409,7 +409,7 @@ func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer, cf
 		return fmt.Errorf("error creating WAL: %w", err)
 	}
 
-	i.writeHandler = remote.NewWriteHandler(i.logger, i.reg, i.wal)
+	i.writeHandler = remote.NewWriteHandler(i.logger, reg, i.wal)
 
 	i.discovery, err = i.newDiscoveryManager(ctx, cfg)
 	if err != nil {

--- a/internal/static/metrics/instance/instance_test.go
+++ b/internal/static/metrics/instance/instance_test.go
@@ -258,7 +258,8 @@ func TestInstance_Recreate(t *testing.T) {
 	cfg.RemoteFlushDeadline = time.Hour
 
 	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
-	inst, err := New(prometheus.NewRegistry(), cfg, walDir, logger)
+	currentReg := prometheus.NewRegistry()
+	inst, err := New(currentReg, cfg, walDir, logger)
 	require.NoError(t, err)
 
 	ctx, cancel := context.WithCancel(context.Background())
@@ -278,6 +279,7 @@ func TestInstance_Recreate(t *testing.T) {
 
 	// Recreate the instance, no panic should happen.
 	require.NotPanics(t, func() {
+		inst, err := New(currentReg, cfg, walDir, logger)
 		require.NoError(t, err)
 		runInstance(t, inst)
 

--- a/internal/static/metrics/instance/instance_test.go
+++ b/internal/static/metrics/instance/instance_test.go
@@ -278,7 +278,6 @@ func TestInstance_Recreate(t *testing.T) {
 
 	// Recreate the instance, no panic should happen.
 	require.NotPanics(t, func() {
-		inst, err := New(prometheus.NewRegistry(), cfg, walDir, logger)
 		require.NoError(t, err)
 		runInstance(t, inst)
 

